File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    private final Set<Integer> primeSet = new HashSet<>();
    
    private void generatePermutations(String current, String numbers, boolean[] visited) {
        if (!current.isEmpty()) {
            int num = Integer.parseInt(current);
            if (isPrime(num)) {
                primeSet.add(num);
            }
        }

        for (int i = 0; i < numbers.length(); i++) {
            if (!visited[i]) {
                visited[i] = true;
                generatePermutations(current + numbers.charAt(i), numbers, visited);
                visited[i] = false;
            }
        }
    }

    private boolean isPrime(int number) {
        if (number < 2) return false;
        if (number == 2) return true;
        if (number % 2 == 0) return false;

        for (int i = 3; i <= Math.sqrt(number); i += 2) {
            if (number % i == 0) return false;
        }

        return true;
    }
    
    public int solution(String numbers) {
        boolean[] visited = new boolean[numbers.length()];
        generatePermutations("", numbers, visited);
        return primeSet.size();
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations


def solution(numbers):
    result = set()
    
    for i in range(1, len(numbers) + 1):
        for x in permutations(numbers, i):
            number = int(''.join(x))
            result.add(number)
    
    return sum(is_prime_number(x) for x in result)


def is_prime_number(x):
    if x <= 1:
        return False
    
    for i in range(2, int(x ** 0.5) + 1):
        if x % i == 0:
            return False
        
    return True
